[pigeon] Added basic C++ support

azchohfi opened the pull request:
This PR adds basic C++ support for Pigeon. I've used the video_player pigeon file as a minimum bar for this, so there are things that are known to not work at all, like async. This code was based on a mix of the Obj-C generator and the Java generator.
I need to add more tests. I've added tests only to the new parameters, and not to the generator itself.
We can work together on getting a min set of tests for this before we merge it, but I wanted to share what I've came up with so far to reduce re-work.

Fixes https://github.com/flutter/flutter/issues/70310

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

stuartmorgan commented:
Thanks for the contribution!

@gaaclarke Can you work with @azchohfi on figuring out the right set of tests, what the missing coverage is, README updates, etc. to get this to a land-able state?

gaaclarke commented:
Here's what we should have for this, a new `flutter/platform_tests` test for c++, something like `flutter/platform_tests/cpp_tests`.  It will have a test runner (gtest?).  `run_tests.py` will use the new generator to generate c++ code that will get linked into the test runner and run unit tests on the generated c++ code.

Something like:
```
flutter/
  ↳ platform_tests/
    ↳ cpp_tests/
      ↳ CMakeLists.txt
      ↳ src/
        ↳ simple_test.cc
        ↳ generated/
          ↳ some_generated_cc_file_from_pigeon.cc
          ↳ some_generated_cc_file_from_pigeon.h
```

If we go down that route we'll have to specify the location of a flutter library to link, probably a desktop one.  I haven't played with the c++ platform channels much to know what makes the most sense.  We could opt to run the test runner on an iOS simulator or Android emulator if that makes things easier.

Let me know what you think of that.  I can get the basic outline working if need be.

stuartmorgan commented:
> It will have a test runner (gtest?).

GoogleTest is what we're using for Windows plugin unit tests now, yes.

> If we go down that route we'll have to specify the location of a flutter library to link, probably a desktop one.

Specifically the Windows desktop library; it's the only one that uses the C++ channel code.

It looks like you're using a dummy app project purpose for this for iOS and Android? We should be able to do the same here. (Although it's not clear to me why these are being run in a custom harness using a fake app project rather than just being set up as plugin native unit tests using a partially-generated plugin.)

> We could opt to run the test runner on an iOS simulator or Android emulator if that makes things easier.

Why would we want to run on an emulator instead of on the host?

gaaclarke commented:
> Specifically the Windows desktop library; it's the only one that uses the C++ channel code.

Is it the only engine that includes the c++ channels?

> It looks like you're using a dummy app project purpose for this for iOS and Android? We should be able to do the same here.

Yep, that's possible.  It's a bit more complicated though because we don't have a windows CI bot yet and all the test automation is in shell scripts.

If they develop it with CMake, it should be easier to write something that can be run locally on windows and run in the existing ci bots.

> Why would we want to run on an emulator instead of on the host?

My thought is that we have existing infrastructure for running tests on android and iOS.  For someone developing on windows it might be easier to latch into that.

stuartmorgan commented:
> > Specifically the Windows desktop library; it's the only one that uses the C++ channel code.
> 
> Is it the only engine that includes the c++ channels?

Yes. It wraps C APIs that only the Windows embedding (of the official embeddings) has, so it would be useless in any of our other embeddings.

> > It looks like you're using a dummy app project purpose for this for iOS and Android? We should be able to do the same here.
> 
> Yep, that's possible. It's a bit more complicated though because we don't have a windows CI bot yet and all the test automation is in shell scripts.

We can get Windows LUCI bots set up here any time using the same structure we're now using for flutter/plugins. I can get that started tomorrow.

> If they develop it with CMake, it should be easier to write something that can be run locally on windows and run in the existing ci bots.

Our CI should build and run in the same environment users would be building and running in, which is Windows. Minor compiler differences are still a very real issue for C++.

> > Why would we want to run on an emulator instead of on the host?
> 
> My thought is that we have existing infrastructure for running tests on android and iOS. For someone developing on windows it might be easier to latch into that.

Between my usual refrain of "we shouldn't continue down the road of making Pigeon tests increasingly bespoke" and the issue above (we should test what users will run; zero people will run the generated code on iOS or Android) I'm strongly against that idea. Building and running a GoogleTest binary on the host is trivial, and if we set this test up as an actual plugin (which I think is how all of these tests should work), we already have tooling support for running it in the repo. We just need a step before it in the CI script to do the generation.

gaaclarke commented:
I played around with flutter windows a bit more to get a better idea.  The layout for the tests will be more like this:
```
flutter/
  ↳ platform_tests/
    ↳ cpp_tests/
      ↳ pubspec.yaml
      ↳ windows/
        ↳ CMakeLists.txt
        ↳ runner/
          ↳ simple_test.cc
          ↳ generated/
            ↳ some_generated_cc_file_from_pigeon.cc
            ↳ some_generated_cc_file_from_pigeon.h
```
Where `flutter/platform_tests/cpp_tests` was created with `flutter create --platforms window` and the `CMakeLists.txt`s are edited to add a test runner target.  I guess the easiest thing would be to check in the gtest library in the repository to link in.  Potentially we could add a download step the test harness once we have a better idea what that will look like.

azchohfi commented:
Thanks for the help @gaaclarke and @stuartmorgan.
Since this is my very first contribution to Flutter, I'm still not sure on regards to where things should be.

1) @gaaclarke, when you say `flutter/platform_tests/cpp_tests`, are you saying that I should send another PR to https://github.com/flutter/platform_tests with the tests, or that I should be creating a cpp_tests folder here in the pigeon package, similar to https://github.com/flutter/packages/tree/master/packages/pigeon/platform_tests/android_unit_tests? I'm leaning towards the latter, but I'm not sure.

2) I never did any GTest, but I believe I can use this commit as a reference:
https://github.com/flutter/plugins/commit/90fd90ed62571fb765df8c49000b14e3c8c643ec
@stuartmorgan, let me know if you have a better example.

3) Also, I would probably need some help with the CI scripts to run these in a Windows machine, as I also don't know where they live, or how they work.

stuartmorgan commented:
> I guess the easiest thing would be to check in the gtest library in the repository to link in.

We could, but for a one-off I would use the same structure we're currently using for plugin tests where it's pulled by the CMake setup.

stuartmorgan commented:
> 2. I never did any GTest, but I believe I can use this commit as a reference:
>    [flutter/plugins@90fd90e](https://github.com/flutter/plugins/commit/90fd90ed62571fb765df8c49000b14e3c8c643ec)
>    @stuartmorgan, let me know if you have a better example.

That's the best reference. The caveat is that it's a plugin, not an app—thus my comments above about potentially changing the way Pigeon native tests are structured. But if @gaaclarke is committed to the app-based version, it shouldn't be hard to adapt.

> 3. Also, I would probably need some help with the CI scripts to run these in a Windows machine, as I also don't know where they live, or how they work.

Currently they don't :) I'm starting the process of getting those configured for this repo though; hopefully it'll be up and running by sometime next week. Because of the way Pigeon tests are currently structured though, ideally the CI would just call Pigeon's run_tests.sh and that would do everything that's needed.

(Alternately, this could be an opportunity to start the process of moving Pigeon away from a complex local script, and toward standard repository tooling; I can discuss with @gaaclarke offline and we can follow up here.)

stuartmorgan commented:
The Windows CI support just landed, so it should be possible to add tests for this now; you'll just need to merge in the latest master.

The tooling changes to move to running the tests here with the standard plugin tooling are less trivial than I initially thought, since they don't run sub-packages (and adding that without unintended effects could be a bit tricky). I think a hybrid is the easiest way to move forward. We could:
- land a dummy plugin project that includes Windows
- incorporate the unit test structure that's being adopted in our 1P plugins (since it's not yet in the template)
- make a script—either via additions to run.sh (our Windows CI has bash) or a new Dart script to make it easier to run locally on Windows—that generates the test files into that plugin, and runs the tests.

If/when we later add the necessary support to `flutter_plugin_tools` to run the tests, we could eliminate that part of the script; the rest would continue to work as-is.

gaaclarke commented:
The tests are hooked up now and can be run now with `dart run pigeon:run_tests -t windows_unittests`.

gaaclarke commented:
I looked a bit more into the output this generates now.  One thing we should fix is that the generated classes have public methods for encoding and decoding to maps.  We don't make those public anywhere else and want to avoid advertising that since we want the ability to change the wire format to something else.  On Dart the encode/decode methods are publicly available but the actual data those create and use is opaque.  So either we make the methods not public, or if they stay public we obfuscate the actual datatype they are using.

gaaclarke commented:
Sorry, I'm just typing up things as I notice them.

1) We should switch the signature for the `Setup` functions to take in a unique_ptr to the api.  Then inside of the `Setup` function we can convert it to a shared_ptr that gets captured by all of the lambda that are passed into `SetMessageHandler`, that way the api instance will live for the duration of the channel.
1) Not a huge deal for MVP but there are a few instances where we should be reference types, like when catching exceptions.
1) Some of the generated code variable names can be cleaned up, like `encodableresult`.

azchohfi commented:
I need to work on this again. My bad for leaving this to rot.

azchohfi commented:
@gaaclarke do you happen to know why the local_tests are failing? The crash I had seems related to iOS/MacOS, which I have not touched. Also, this seems to be running the ./packages/pigeon/run_tests.sh, which seems to be deprecated, and running run_tests.dart locally works fine for me.

azchohfi commented:
Oh, is this related to this?
https://github.com/flutter/packages/blob/0f153902959efa165fd8d136061677510055185c/packages/pigeon/bin/run_tests.dart#L313

azchohfi commented:
Yeah, still failing with a timeout.

azchohfi commented:
Seems to be an issue on the master branch of Flutter. I'll wait.

gaaclarke commented:
(I'm getting the timeout too.  I'll poke around tomorrow to see what is going on: https://github.com/flutter/flutter/issues/98555)

gaaclarke commented:
If you rebase the timeout should be gone now.

azchohfi commented:
@gaaclarke anything else needed here?

gaaclarke commented:
We just need some documentation that says what the current status of support is.  That means probably somewhere in the README.md, maybe a tiny note in `-h`.

azchohfi commented:
Ok, I've tested this more, and the codec was actually required for many scenarios, so I finally did the work and it should now be working quite well. I've migrated https://github.com/gaaclarke/pigeon_plugin_example and it is working perfectly.

azchohfi commented:
@gaaclarke Just fixed the tests here (very minor mistake). Let me know if you still want something on pigeon's help. I've added docs for the new parameters, but I didn't find anything specific to what is/is not supported on each platform on the help (-h). Let me know if you still want anything there.
Also, I would like to add some more tests in the future with more complex scenarios (like async, etc), but I believe this is good enough for now.

gaaclarke commented:
> Also, I would like to add some more tests in the future with more complex scenarios (like async, etc), but I believe this is good enough for now.

Yea, that's what I was just looking at too.  I just added a unit test for the cpp_generator so that people have no excuse to add them in the future.  I want to land this too but since the test coverage is small we might have to designate this as "experimental" for now.  Once this lands, it's my responsibility to maintain so it would be nice to have a backstop considering the test coverage.

I'm about to land a contributors PR for swift support, too.  We might have to come up with some compatibility grid on the README so people know what works and what doesn't.  That can also guide contributors.

gaaclarke commented:
To be clear I'm not asking for any more changes.  I'm just looking this over.  I might make some tweaks to the wording of the README and backfill some tests this afternoon =)

azchohfi commented:
This is great! I'm testing this against the video_player (I have a semi-working branch of it), and I'll send fixes if I find anything.
I believe the one thing that does not work properly yet is an API method with multiple parameters.

gaaclarke commented:
I added a compilation test (where code is generated and compiled but didn't test that it runs correctly).  This is a cheap and easy way to get some more coverage.

gaaclarke commented:
@azchohfi if I add `pigeons/async.dart` I get compilation errors.

diff:
```diff
--- a/packages/pigeon/bin/run_tests.dart
+++ b/packages/pigeon/bin/run_tests.dart
@@ -271,6 +271,15 @@ Future<int> _runWindowsUnitTests() async {
     return generateCode;
   }
 
+  generateCode = await _runPigeon(
+    input: './pigeons/async_handlers.dart',
+    cppHeaderOut: '$windowsUnitTestsPath/windows/test/async_handlers.g.h',
+    cppSourceOut: '$windowsUnitTestsPath/windows/test/async_handlers.g.cpp',
+  );
+  if (generateCode != 0) {
+    return generateCode;
+  }
+
   final Process compile = await _streamOutput(Process.start(
       'flutter', <String>['build', 'windows', '--debug'],
       workingDirectory: '$windowsUnitTestsPath/example', runInShell: true));
--- a/packages/pigeon/platform_tests/windows_unit_tests/windows/CMakeLists.txt
+++ b/packages/pigeon/platform_tests/windows_unit_tests/windows/CMakeLists.txt
@@ -55,6 +55,8 @@ add_executable(${TEST_RUNNER}
   test/pigeon_test.cpp
   test/all_datatypes.g.cpp
   test/all_datatypes.g.h
+  test/async_handlers.g.cpp
+  test/async_handlers.g.h
   test/message.g.cpp
   test/message.g.h
   ${PLUGIN_SOURCES}
```

errors:
```
C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\windows\flutter\ephemeral\cpp_client_wrapper\include\flutter/basic_message_channel.h(22,53): error C2182: 'reply': illegal use of type 'void' [C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\build\windows\plugins\windows_unit_tests\windows_unit_tests_test.vcxproj]
C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\windows\flutter\ephemeral\.plugin_symlinks\windows_unit_tests\windows\test\async_handlers.g.cpp(81,75): error C2663: 'std::_Tree<std::_Tmap_traits<_Kty,_Ty,_Pr,_Alloc,false>>::insert': 8 overloads have no legal conversion for 'this' pointer [C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\build\windows\plugins\windows_unit_tests\windows_unit_tests_test.vcxproj]
C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\windows\flutter\ephemeral\.plugin_symlinks\windows_unit_tests\windows\test\async_handlers.g.cpp(81,1): error C2664: 'void Api2Host::calculate(Value,std::function<void (const Value &)>)': cannot convert argument 2 from 'Api2Host::Setup::<lambda_a31298d1458cada3e41215767b43f015>::()::<lambda_03de75a11f3b0b04ff776ab225165e07>' to 'std::function<void (const Value &)>' [C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\build\windows\plugins\windows_unit_tests\windows_unit_tests_test.vcxproj]
C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\windows\flutter\ephemeral\.plugin_symlinks\windows_unit_tests\windows\test\async_handlers.g.cpp(76,14): error C2664: 'void flutter::BasicMessageChannel<flutter::EncodableValue>::SetMessageHandler(const std::function<void (const T &,const std::function<void (const T &)> &)> &) const': cannot convert argument 1 from 'Api2Host::Setup::<lambda_a31298d1458cada3e41215767b43f015>' to 'const std::function<void (const T &,const std::function<void (const T &)> &)> &' [C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\build\windows\plugins\windows_unit_tests\windows_unit_tests_test.vcxproj]
C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\windows\flutter\ephemeral\.plugin_symlinks\windows_unit_tests\windows\test\async_handlers.g.cpp(99,64): error C2663: 'std::_Tree<std::_Tmap_traits<_Kty,_Ty,_Pr,_Alloc,false>>::insert': 8 overloads have no legal conversion for 'this' pointer [C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\build\windows\plugins\windows_unit_tests\windows_unit_tests_test.vcxproj]
C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\windows\flutter\ephemeral\.plugin_symlinks\windows_unit_tests\windows\test\async_handlers.g.cpp(99,1): error C2664: 'void Api2Host::voidVoid(std::function<void (int &)>)': cannot convert argument 1 from 'Api2Host::Setup::<lambda_1a5d21da010418dcba2846db68cbe451>::()::<lambda_50572e48b39fa1754657350c229c2d05>' to 'std::function<void (int &)>' [C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\build\windows\plugins\windows_unit_tests\windows_unit_tests_test.vcxproj]
C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\windows\flutter\ephemeral\.plugin_symlinks\windows_unit_tests\windows\test\async_handlers.g.cpp(96,14): error C2664: 'void flutter::BasicMessageChannel<flutter::EncodableValue>::SetMessageHandler(const std::function<void (const T &,const std::function<void (const T &)> &)> &) const': cannot convert argument 1 from 'Api2Host::Setup::<lambda_1a5d21da010418dcba2846db68cbe451>' to 'const std::function<void (const T &,const std::function<void (const T &)> &)> &' [C:\Users\gaaclarke\dev\packages\packages\pigeon\platform_tests\windows_unit_tests\example\build\windows\plugins\windows_unit_tests\windows_unit_tests_test.vcxproj]
Build process failed.
```

One of the errors is inserting into the `EncodableMap` and the other is a lambda that can't be cast to a `MessageReply<void>`.

I gave it a look but I don't have much experience with the C++ interface, can you check it out please?

azchohfi commented:
Working on it.
I've migrated the video_player to use the latest pigeon, and I had to do some adjustments, but I've found an issue in decoding already, so this is good progress. Will also work on the async.

azchohfi commented:
@gaaclarke I've added your diff and tested it locally. I fixed it both for the video_player pigeon, as well as the async one. It should support multiple parameters now.

azchohfi commented:
I'll try to enable all of them (interection between android and ios ones).

gaaclarke commented:
Sounds good, I was going to try that.  Not everything needs to work, but `messages.dart` `all_datatypes.dart` and `async.dart` are important considering where we said the code is at.  It might be a good way to verify what we don't support yet too, like `nullable_returns.dart` and `primitive.dart`.  Sorry, those files are a bit messy.  I haven't come up with a good system for organizing them yet.

gaaclarke commented:
We should have one test that actually executes the code too.  I thought we had one but looking at it more closely it is just using method channels, not pigeon.  I'll give it a shot.

azchohfi commented:
Ok, now multiple parameters should be working, and all the tests are passing, which is great! I've tested this version on both the video_player, as well as in pigeon_plugin_example, and everything seems to be working quite well.

gaaclarke commented:
@azchohfi I added a test that will exercise some of the generated code.  We should probably make one that is encoding and decoding data.

gaaclarke commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,670 @@
+              'void ${_makeSetter(field)}(${hostDatatype.datatype} setterArg);');
+          indent.addln('');
+        }
+        indent.write('flutter::EncodableMap ToEncodableMap(); ');
```
We need to make these methods private.  We don't want people to rely on them or call them because we want to completely abstract away the wire format of pigeon.  I think the easiest thing is to make these private and make the serializer a `friend`.

gaaclarke replied:
Done.  I think we might be able to make something a bit cleaner if we just put the encode and decode functions in the .cpp file and don't include them in the header, but this gets is what we want for now.

gaaclarke commented:
Another update:
1) Added a test that shows sending an argument
1) Made the codec's `writeValue` method public to match its superclass
1) Made the codec's singleton getter match the naming convention of its superclass

gaaclarke reviewed (approved):
LGTM.  I'm feeling better now that we have more compilation tests and some tests that exercise the code.
![high five](https://media1.giphy.com/media/26BkMmfVnpxPcz4kg/giphy.gif?cid=790b76115a58c3c975608d13f135c2890ca56ffbce3d283f&rid=giphy.gif&ct=g)

gaaclarke commented:
Adding @stuartmorgan for secondary review.

azchohfi commented:
Awesome! Thanks for all your help on getting this ready to merge. Really appreciate it.

stuartmorgan commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+      for (final Method method in api.methods) {
+        final String returnType = method.isAsynchronous
+            ? 'void'
+            : _cppTypeForDartType(method.returnType);
```
I'm concerned about the structure here; you've eliminated the ability to do the equivalent of `result->Error` in a non-Pigeon plugin, which is important functionality. AFAIK the only way to return an error would be to throw which a) we should absolutely not be encouraging/requiring people to do as a normal error handling mechanism in C++, and b) provides less information than a usual method channel error (code + description + arbitrary details).

We need to decide on a structure that allows for errors. Maybe that's something like a generated `ErrorOr<SomeReturnType>` to use as the return value?

azchohfi replied:
I did a similar implementation to the Java one. The one info that I'm not wrapping is the stack trace, but if the developer throws an exception, it will be handled and wrapped properly. The iOS implementation has a FlutterError type, so we could generate a generic equivalent of that, like your ErrorOr<T>.

azchohfi replied:
Ok, this took me a while, but this seems to be working quite well.

stuartmorgan commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+                    indent.writeln(
+                        'throw exception("$argName unexpectedly null.");');
+                  });
+                  */
```
Why is there commented-out code?

azchohfi replied:
This is for nullables, which is not implemented here yet. I've tried with optional<T>, but c++17 doesn't support optional references, so I think I'll go the simpler pointer route. This will then be uncommented.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+
+          if (func.returnType.baseName == 'int') {
+            indent.format('''
+if(const int32_t* pval$output = std::get_if<int32_t>(&args))
```
`pval` here was for pointers. I'll restructure the code and make it `pointer_`

azchohfi replied:
Done.

azchohfi commented on packages/pigeon/CHANGELOG.md:
```diff
@@ -1,3 +1,6 @@
+## 1.1.0
+* [c++] Added C++ code generator.
```
Done.

azchohfi commented on packages/pigeon/README.md:
```diff
@@ -5,8 +5,9 @@ host platform type-safe, easier and faster.
-Currently Pigeon only supports generating Objective-C code for usage on iOS and
-Java code for Android.  The Objective-C code is
+Currently Pigeon supports generating Objective-C code for usage on iOS, Java
+code for Android, and has experimental support for C++ Windows. The Objective-C
```
Done.

azchohfi commented on packages/pigeon/README.md:
```diff
@@ -35,16 +37,36 @@ doesn't need to worry about conflicting versions of Pigeon.
+   to your `windows/CMakeLists.txt` file.
+1) Implement the generated C++ abstract class for handling the calls on Windows,
+   set it up as the handler for the messages.
 1) Call the generated Dart methods.
```
Done.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+  indent.writeln('// $generatedCodeWarning');
+  indent.writeln('// $seeAlsoWarning');
+  indent.addln('');
+  indent.writeln('#pragma once');
```
Done.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+  indent.writeln('#include <flutter/basic_message_channel.h>');
+  indent.writeln('#include <flutter/binary_messenger.h>');
+  indent.writeln('#include <flutter/standard_message_codec.h>');
+  indent.writeln('#include <map>');
```
Done.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+  assert(api.location == ApiLocation.host);
+
+  indent.writeln(
+      '/** Generated class from Pigeon that represents a handler of messages from Flutter.*/');
```
Done.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+          indent.writeln('callback();');
+        } else {
+          indent.writeln(
+              'auto decodedReply = GetCodec().DecodeMessage(reply, reply_size);');
```
I've reviewed this and I believe it is much better now. I'll resolve this conversation, but let me know if you want to change anything else.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+          'auto channel = std::make_unique<flutter::BasicMessageChannel<flutter::EncodableValue>>(');
+      indent.inc();
+      indent.inc();
+      indent.writeln('binaryMessenger, "$channelName", &GetCodec());');
```
Done.

azchohfi commented on packages/pigeon/README.md:
```diff
@@ -118,6 +140,16 @@ public interface Api2Host {
+/** Generated class from Pigeon that represents a handler of messages from Flutter.*/
+class Api2Host {
+public:
+    virtual void calculate(Value value, flutter::MessageReply<Value> result) = 0;
```
Done. I believe I got it all with a protected default ctr but let me know if we need anything else.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+              'void ${_makeSetter(field)}(${hostDatatype.datatype} setterArg);');
+          indent.addln('');
+        }
+        indent.writeln('${klass.name}();');
```
Done.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+          indent.writeln('${hostDatatype.datatype} ${field.name};');
+        }
+      });
+    });
```
Done. I've added some other extra blank lines for improved reading.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+        for (final EnumeratedClass customClass in getCodecClasses(api, root)) {
+          indent
+              .write('if(customValue->type() == typeid(${customClass.name}))');
+          indent.scoped('{', '} else ', () {
```
I've refactored this so each item returns at the end of their own if statements. Unfortunately, we can't use switch statements for typeids.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+\t$output = *pval2$output;''');
+          } else if (!isBuiltin) {
+            indent.writeln(
+                'if(const flutter::EncodableMap* pval$output = std::get_if<flutter::EncodableMap>(&args)) $output = $returnType(*pval$output);');
```
Improved this at a bunch of different places.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+      indent.scoped('{', '}', () {
+        for (final NamedType field in klass.fields) {
+          indent.writeln(
+              'auto encodable${field.name} = map.at(flutter::EncodableValue("${field.name}"));');
```
Done.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+  indent.writeln(
+      '/** Sets up an instance of `${api.name}` to handle messages through the `binaryMessenger`. */');
+  indent.write(
+      'void ${api.name}::Setup(flutter::BinaryMessenger* binaryMessenger, ${api.name}* api) ');
```
Done.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+        'void $codecName::WriteValue(const flutter::EncodableValue& value, flutter::ByteStreamWriter* stream) const');
+    indent.writeScoped('{', '}', () {
+      indent.write(
+          'if(const flutter::CustomEncodableValue* customValue = std::get_if<flutter::CustomEncodableValue>(&value))');
```
Done.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,683 @@
+      sendArgument = 'flutter::EncodableValue()';
+    } else {
+      final Iterable<String> argTypes =
+          func.arguments.map((NamedType e) => _cppTypeForDartType(e.type));
```
I've updated this to use const ref for non-built-in types. Should we use std::optional to allow nullables? Maybe this would require C++20 for references, so I'm unsure about the right approach here.

azchohfi replied:
Ok, I gave it a shot, but we need more e2e tests to properly test this.

azchohfi commented:
I had to create a special case for void returns (it returns an `std::optional<FlutterError>`), so all methods are either `ErrorOr<T>` or `std::optional<FlutterError>`.
This is the generated class:
```c++
class FlutterError {
 public:
...
  std::string code;
  std::string message;
  std::string details;
};
template<class T> class ErrorOr {
  std::variant<T, FlutterError> v;
  bool ok = true;
 public:
  ErrorOr() { new(&v) T(); }
  ErrorOr(const T& rhs) { new(&v) T(rhs); }
  ErrorOr(const FlutterError& rhs) : ok(false) {
    new(&v) FlutterError(rhs);
  }
  bool hasError() const { return !ok; }
  const T& value() const { return std::get<T>(v); };
  const FlutterError& error() const { return std::get<FlutterError>(v); };
};
```

gaaclarke commented:
Looks like @azchohfi marked all of the feedback resolved.  Rerequesting review from @stuartmorgan.

azchohfi commented:
Ping @stuartmorgan

xiaowei-guan commented:
@azchohfi Hi, when I try use this PR to generate C++ code,  I meet a build error:
  src/messages.cc:575:19: error: no matching conversion for functional-style cast from 'const char [27]' to 'std::exception'
                       throw std::exception("msg_arg unexpectedly null.");

azchohfi commented:
@xiaowei-guan could you provide more details? I'm not sure why this is happening.
Aside from this, @stuartmorgan anything else blocking this PR?

stuartmorgan commented:
> Aside from this, @stuartmorgan anything else blocking this PR?

Currently it's me having time to do another full review of the updated PR and generated code. I've been OOO part of the time, and it's also a large context switch from my normal work to review a C++ API generator, so it's been hard to schedule in. It's still on my radar though.

xiaowei-guan commented:
> Currently it's me having time to do another full review of the updated PR and generated code. I've been OOO part of the time, and it's also a large context switch from my normal work to review a C++ API generator, so it's been hard to schedule in. It's still on my radar though.



> @xiaowei-guan could you provide more details? I'm not sure why this is happening. Aside from this, @stuartmorgan anything else blocking this PR?

I find std::exception don't have a constructor that takes a string atgument at my side. Does this PR only support windows?

stuartmorgan commented:
> Does this PR only support windows?

Yes. Pigeon generators interact with the plugin APIs, so are inherently platform-specific.

What platform were you trying to use it for? There are no other official embeddings that have a C++ plugin API layer.

xiaowei-guan commented:
> ou trying to use it for? There are no other official embeddings that have a C++ plugin API layer.

Ok, I got it, we are using Tizen platform(https://github.com/flutter-tizen), we develop plugin based on C++ plugin API layer.
I meet some build errors when generating C++ code by Pigeon. If only support windows, I need modify it for Tizen platfrom.

stuartmorgan commented:
> Ok, I got it, we are using Tizen platform(https://github.com/flutter-tizen), we develop plugin based on C++ plugin API layer.

I see; if you're using the engine's shared C++ code (meaning this is a fork of the engine, presumably?), then the same generator should mostly work.

> I meet some build errors when generating C++ code by Pigeon. If only support windows, I need modify it for Tizen platfrom.

In general, the expectation is that every embedding would need its own Pigeon generator because every embedding has its own API surface. If you're using the exact same API surface as windows, that's a very special case—forking Pigeon the way you would have had to fork the engine is probably your best option. It should be a fairly minimal fork.

stuartmorgan commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,794 @@
+      indent.writeln('virtual ~${api.name}() { };');
+      for (final Method method in api.methods) {
+        final String returnTypeName = method.returnType.isVoid
+            ? 'std::optional<FlutterError>'
```
I'm wondering about the ergonomics of this; it's essentially a double-nested `variant` to express three states. Should we instead have a `NullableErrorOr` that can directly express the three states? It seems (naively; I haven't tried implementing an API using this yet, so maybe you've found that it's not the case) that it would be easier to author implementations that way.

azchohfi replied:
I wanted to leverage what was already implemented by std as much as possible, so this seemed like the best approach.

stuartmorgan reviewed (changes requested):
I think this is getting close at a high level, with the caveat that it's a ton of generated code and I'm not deeply familiar with all the test cases so this isn't a completely thorough review of whether all the generated code matches expectations. Perhaps @gaaclarke should do another sweep of that once the next round of changes is in. The main thing I reviewed for here was places where it looked like the reference/value or optional/non-optional semantics didn't match what I was expecting.

The `ErrorOr` approach looks like it works out pretty well (although we'll have to see what it feels like to write against it). Do you have any plugins actually using this iteration? It would be interesting to see a real plugin written against it; we've found a lot of things in the other generators that way. If not, we should just assume we'll need to iterate more on it once we actually try it out in practice.

I'm just going to do some of this as a high-level comment rather than try to link comments back to the generating lines, since I'm just going through `platform_tests\windows_unit_tests\windows\test\`.

- `all_datatypes` is still a sea of pass-by-value:
  - FlutterError arguments are all PBV strings instead of const refs
  - `Everything` is both passing and returning everything by value; only `bool`, `int*`, and `double` should be. Collections shouldn't be value-based.
- In `nullable_returns`:
  - the (incorrectly named; it returns an `int?`) `NonNullHostApi` `doIt` is returning an `ErrorOr<int64_t>`, so can't actually return null.
  - `NullableArgFlutterApi` has a reply object that is a `const int64_t&`; int types should be passed by value.
- Relatedly, in `primitive`, all the non-complex types are incorrectly being passed by reference.
- In `list`, `testList` is a `List?` in the Dart specification, but the C++ version wouldn't allow null.

stuartmorgan commented:
@gaaclarke Have you had any more thoughts about how exactly you want to handle the experimental nature of this?

Since we plan to make it experimental at first anyway, we could probably call this PR done on the next iteration as long as we're able to make breaking changes to Windows easily, then do further improvements/fixes as follow-ups once we find them in actual use, rather than trying to get everything right up-front. This iteration's structure feels close enough to what I expect a final version to look like that I think the next version would make a very reasonable stopping point for the initial landing.

azchohfi commented:
@stuartmorgan thanks for the review. They make a lot of sense and I will go through all of these points soon, just likely not this week. In the meantime, I did update the video_player plugin to build for windows. It is a partial implementation, as there is no video output just yet (audio works fine). It is implemented using Windows' Media Foundation APIs:
https://github.com/azchohfi/plugins/tree/video_player_windows/packages/video_player/video_player_windows/windows

I've already rebased this on the latest, but if you want to run this, just be careful since it is pointing to absolute paths for pigeon, where I was doing the experimentation.

stuartmorgan commented:
Thanks, it's helpful to see it in a real example! It does highlight a downside to the error-as-return value approach, which is all of the `return {};` statements, but unfortunately I don't see a better solution unfortunately (since exceptions aren't an option, and an out-param would be worse).

One small future improvement I can see from that code—not something that you should feel like you should add to this PR, since it's just a very minor usability improvement—would be generating a full-argument constructor for message objects so that:
```
    TextureMessage textureMessage;
    textureMessage.setTextureId(textureId);
    return textureMessage;
```

could just be `return TextureMessage(textureId)`.

(But then, now that Pigeon has multi-argument support, we shouldn't actually need most of those message objects at all, making it even more of a minor point.)

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,794 @@
+\t\t: code(arg_code), message(arg_message), details(arg_details) {};
+\tstd::string code;
+\tstd::string message;
+\tstd::string details;
```
I could change this to EncodableValue, but I just copied what the Java implementation does, and we only ever send strings, so I didn't see the point. I wanted to be as specific as possible.

stuartmorgan replied:
Since Pigeon is intended to be general-purpose, I'd rather we not restrict what people can provide as errors. There are scenarios where sending objects back in details is useful.

gaaclarke commented:
@stuartmorgan and I talked offline the namespace named `experimental` is good for C++ but doesn't scale well across other languages.  It's best just to add the `experimental` prefix to the flag.  We can just alias it once we remove it so no one is broken.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,794 @@
+                    indent.write('if ($encodableArgName.IsNull()) ');
+                    indent.scoped('{', '}', () {
+                      indent.writeln(
+                          'throw std::exception("$argName unexpectedly null.");');
```
Fixed.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,794 @@
+    if (getCodecClasses(api, root).isNotEmpty) {
+      indent.writeScoped(' public:', '', () {
+        indent.writeln(
+            'void WriteValue(const flutter::EncodableValue& value, flutter::ByteStreamWriter* stream) const;');
```
Fixed.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,794 @@
+};
+template<class T> class ErrorOr {
+\tstd::variant<T, FlutterError> v;
+\tbool ok = true;
```
Fixed.

azchohfi commented:
@stuartmorgan I'm struggling with ErrorOr since it uses variant and it can't hold a reference type. Any suggestions here? Not sure what the best solution would be.

azchohfi commented:
@stuartmorgan let me know what you think. I'm just passing by value in this case.
I think I got most, if not all of the feedback you provided in your last review.

azchohfi commented:
@stuartmorgan I've also updated https://github.com/azchohfi/plugins/tree/video_player_windows/packages/video_player/video_player_windows/windows
with this.

stuartmorgan commented:
> @stuartmorgan I'm struggling with ErrorOr since it uses variant and it can't hold a reference type. Any suggestions here? Not sure what the best solution would be.

Since copying it could be expensive, depending on what's inside it, I would pass a `unique_ptr` instead so that there's a clear ownership transfer without copies. (In theory people could get the same effect by `std::move`ing the pass-by-value version, but it would be very easy for people to not think of that and get poor performance, so I would prefer to enforce a lack of copy.)

stuartmorgan commented:
(I'll review the rest again soon.)

azchohfi commented:
@gaaclarke I'm not sure why the CallSearch test mocks are leaking. Do you happen to know?

gaaclarke commented:
> @gaaclarke I'm not sure why the CallSearch test mocks are leaking. Do you happen to know?

What indication do you have that there is a leak?  There is no dynamic memory allocation in that test, everything is on the stack so it isn't an error in the test.  I'm not that familiar with the C++ API.  You might try adding `Api::SetUp(&mock_messenger, nullptr);`  I'd also look up to see what type "flutter::BinaryMessageHandler" actually is to make sure it is a RAII object.

azchohfi commented:
@gaaclarke I'm seeing this when running the tests locally:
```
> flutter pub run pigeon:run_tests
# Running windows_unittests

💪 Building with sound null safety 💪                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Building Windows application...
Running main() from C:\Users\alzollin\source\repos\flutter_packages\packages\pigeon\platform_tests\windows_unit_tests\example\build\windows\_deps\googletest-src\googletest\src\gtest_main.cc
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from PigeonTests
[ RUN      ] PigeonTests.Placeholder
[       OK ] PigeonTests.Placeholder (1 ms)
[ RUN      ] PigeonTests.CallInitialize
[       OK ] PigeonTests.CallInitialize (0 ms)
[ RUN      ] PigeonTests.CallSearch
unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
[  FAILED  ] PigeonTests.CallSearch (0 ms)
[----------] 3 tests from PigeonTests (2 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test suite ran. (3 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] PigeonTests.CallSearch

 1 FAILED TEST

C:\Users\alzollin\source\repos\flutter_packages\packages\pigeon\platform_tests\windows_unit_tests\example\windows\flutter\ephemeral\.plugin_symlinks\windows_unit_tests\windows\test\pigeon_test.cpp(123): ERROR: this mock object (used in test PigeonTests.CallSearch) should be deleted but never is. Its address is @00000024702FEBA0.
C:\Users\alzollin\source\repos\flutter_packages\packages\pigeon\platform_tests\windows_unit_tests\example\windows\flutter\ephemeral\.plugin_symlinks\windows_unit_tests\windows\test\pigeon_test.cpp(129): ERROR: this mock object (used in test PigeonTests.CallSearch) should be deleted but never is. Its address is @00000024702FEC80.
ERROR: 2 leaked mock objects found at program exit. Expectations on a mock object are verified when the object is destructed. Leaking a mock means that its expectations aren't verified, which is usually a test bug. If you really intend to leak a mock, you can suppress this error using testing::Mock::AllowLeak(mock_object), or you may use a fake or stub instead of a mock.
pub finished with exit code 1
```
 Weirdly enough, the CI is not running them.

azchohfi commented:
Oh, I believe this has not yet been implemented.

azchohfi commented:
I looked at the implementation of the native_test_command, and it seems like it only works for plugins that explicitly define support for a platform, which the pigeon package does not. I'm not sure what the right approach should be here, but the native tests are just not being executed. @stuartmorgan / @gaaclarke Any idea?

stuartmorgan commented:
> I looked at the implementation of the native_test_command, and it seems like it only works for plugins that explicitly define support for a platform, which the pigeon package does not. I'm not sure what the right approach should be here, but the native tests are just not being executed. @stuartmorgan / @gaaclarke Any idea?

There was a lot of complexity with trying to make `native-test` run Pigeon native tests, so `custom-test` does it instead.

Except it turns out I forgot that when I switched from the old shell script runner to the Dart runner I needed to move Pigeon's run_tests.dart from bin/ to tool/ (where it should be according to Dart package structure conventions), so Windows CI stopped running Pigeon's tests. I'll have a PR to fix that in a minute.

stuartmorgan commented:
https://github.com/flutter/packages/pull/1549

stuartmorgan commented:
Pigeon tests will now run in CI if you merge in master again; sorry about that regression.

(I also plan to review this PR again tomorrow.)

azchohfi commented:
Just merged it. We'll see how that goes, and thanks for the fix!

azchohfi commented:
Yeah, this is now running the tests, and I'm seeing exactly the same issue locally.

azchohfi commented:
Aha, I found the issue. It was a nre within the generated code. Simple fix.

stuartmorgan reviewed (changes requested):
Looking at `Everything` again, I'm actually not sure we want const references for the setters, vs `unique_ptrs`. All the setters would still be quite expensive for complex structures because internally it has to copy them still. I think that should be a relatively straightforward change in the generator?

Other than that and a few more inline comments below, I think I'm probably at the limit of what I'm going to find in the abstract environment of a code review, so I'm happy to call this approved from my end once the remaining comments are resolved. I'll try to find time to use it in an actual plugin or two once it's landed, and we can iterate based on that (as we've been doing with the other generators).

azchohfi commented:
@stuartmorgan If we use std::unique_ptr, then the classes are not copy constructable, so we can't pass them to CustomEncodableValue to serialization. Is there a way around this?

azchohfi commented:
Also, there seems to be some more work required to make the unique_ptr work with ErrorOr<T> (.value() doesn't work since it needs a special case for unique_ptr since it can't be copied, so I'm wondering if this whole structure should change. Not sure exactly how to proceed here, but I will investigate. Any guidance would be much appreciated.

stuartmorgan commented:
> If we use std::unique_ptr, then the classes are not copy constructable, so we can't pass them to CustomEncodableValue to serialization. Is there a way around this?

Hm, maybe there's a more fundamental limitation in the Windows platform channel API.

Sounds like this would need a lot more investigation, and maybe even engine API changes, for truly efficient complex object handling. Given that, and that very complex objects is probably not a common case, let's go with the references for now, and if we need to make a breaking change in Pigeon someday to change things (maybe after engine changes), we can always do that.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -294,7 +294,7 @@ const flutter::StandardMessageCodec& ${api.name}::GetCodec() {
                   if (isReferenceReturnType && !returnType.isNullable) {
                     result =
-                        'flutter::CustomEncodableValue(*output.value().get())';
+                        'output.value() ? flutter::CustomEncodableValue(*output.value().get()) : flutter::EncodableValue()';
```
Yup. I still need to fix this.

azchohfi replied:
Fixed.

gaaclarke commented:
@azchohfi for the Swift generator PR I added a validation phase that blocks features that exist in (dart, java, objc) but isn't in swift yet: https://github.com/flutter/packages/pull/976/commits/817e9fff5cf5bf938c7f2a4fa95e23a50081fe20  We should do that here too.  Let me know if you want to take a stab at it, otherwise I can do it too.  Also let me know if you already have a list of things we don't support yet.

azchohfi commented on packages/pigeon/lib/cpp_generator.dart:
```diff
@@ -0,0 +1,832 @@
+}
+
+String _getGuardName(String? headerFileName, String? namespace) {
+  String guardName = 'PIGEON_';
```
I believe this can be fixed in the future, since it is very unlikely to be a breaking change.

azchohfi commented:
@gaaclarke The commit you mentioned is very specific to the feature that the Swift one doesn't support. It should be quite trivial to add these validations to the C++ generator, but I would appreciate some help with this, as the initial code was written to support an older version of pigeon, so some features might actually be really missing. I know that I've added some of them given the feedback on the PR (like non-nullable references), but only to a minimal extent (things build, but not enough tests to mark this as not experimental).

gaaclarke commented:
I ran through some of the more recent tests and I think you've implemented nullable returns and nonnull arguments.  I think the only thing that isn't supported yet is non-null fields.  I added a validator for that.

azchohfi commented:
Should we just remove the non_null_fields.dart file from the tests then? The tests are failing because of it now.

gaaclarke commented:
> @gaaclarke Should we just remove the non_null_fields.dart file from the tests then? The tests are failing because of it now.

Yep, sorry I missed that.

azchohfi commented:
@gaaclarke do you want me to remove the file from the tests?

gaaclarke commented:
> @gaaclarke do you want me to remove the file from the tests?

Yea, I'd just comment them out the invocations from the run_tests script with a note.

azchohfi commented:
@gaaclarke anything needed here?

stuartmorgan reviewed (approved):
The changes LGTM, so we're good to land this and then iterate from there.

@azchohfi Thanks for all the great work on this! I'm looking forward to trying it out on `file_selector` to get some hands-on experience with it :)

stuartmorgan commented:
I filed https://github.com/flutter/flutter/issues/104278 to track non-null field support.

stuartmorgan commented:
@azchohfi By the way, if you would like [contributor access](https://github.com/flutter/flutter/wiki/Contributor-access) just let me know; I'd be more than happy to sponsor you.

azchohfi commented:
@stuartmorgan that would be awesome! I appreciate it!

stuartmorgan commented:
Just ping me on [our Discord](https://github.com/flutter/flutter/wiki/Chat) when you get a chance (that's where the mechanics of adding someone are handled by the admins), and I'll start the process!

azchohfi commented:
@stuartmorgan just sent you a friend request. Thanks again.

clarkezone commented:
Great to see this landed, congrats @azchohfi !
